File: crates/calib/examples/laserline_device_session.rs
//! Example: single laserline device calibration (camera + laser plane).

use calib::core::ScheimpflugParams;
use calib::core::SensorModel;
use calib::laserline_device::{run_calibration, LaserlineDeviceConfig, LaserlineDeviceProblem};
use calib::optim::{LaserPlane, LaserlineMeta, LaserlineView};
use calib::prelude::*;

use nalgebra::{Point3, Rotation3, Translation3, Vector3};

fn make_target_points() -> Vec<Pt3> {
    let spacing = 0.03;
    let mut points = Vec::new();
    for i in 0..6 {
        for j in 0..5 {
            points.push(Pt3::new(i as f64 * spacing, j as f64 * spacing, 0.0));
        }
    }
    points
}

fn make_poses() -> Vec<Iso3> {
    vec![
        Iso3::from_parts(
            Translation3::new(0.0, 0.0, 0.5),
            Rotation3::from_euler_angles(0.0, 0.0, 0.0).into(),
        ),
        Iso3::from_parts(
            Translation3::new(0.05, -0.02, 0.55),
            Rotation3::from_euler_angles(0.15, -0.05, 0.0).into(),
        ),
        Iso3::from_parts(
            Translation3::new(-0.04, 0.03, 0.6),
            Rotation3::from_euler_angles(-0.1, 0.08, 0.02).into(),
        ),
        Iso3::from_parts(
            Translation3::new(0.02, 0.06, 0.52),
            Rotation3::from_euler_angles(0.05, 0.12, -0.04).into(),
        ),
        Iso3::from_parts(
            Translation3::new(-0.06, -0.04, 0.58),
            Rotation3::from_euler_angles(-0.12, -0.04, 0.06).into(),
        ),
    ]
}

fn make_laser_plane(poses: &[Iso3]) -> LaserPlane {
    let laser_normal = Vector3::new(0.1, 0.05, 0.99).normalize();
    let target_center = Pt3::new(0.075, 0.06, 0.0);
    let target_center_cam = poses[0].transform_point(&target_center);
    LaserPlane::new(laser_normal, -laser_normal.dot(&target_center_cam.coords))
}

fn laser_pixels_for_view<Sm: SensorModel<f64>>(
    pose: &Iso3,
    plane: &LaserPlane,
    camera: &Camera<f64, Pinhole, BrownConrady5<f64>, Sm, FxFyCxCySkew<f64>>,
) -> Vec<Pt2> {
    let n_c = plane.normal.as_ref();
    let n_t = pose.rotation.inverse_transform_vector(n_c);
    let d_t = n_c.dot(&pose.translation.vector) + plane.distance;

    let dir = Vector3::new(n_t.y, -n_t.x, 0.0);
    let dir_norm = dir.norm();
    if dir_norm < 1e-9 {
        return Vec::new();
    }
    let dir = dir / dir_norm;
    let (x0, y0) = if n_t.x.abs() > n_t.y.abs() {
        (-d_t / n_t.x, 0.0)
    } else {
        (0.0, -d_t / n_t.y)
    };

    let mut pixels = Vec::new();
    for i in 0..40 {
        let s = (i as f64 / 39.0) * 0.2 - 0.1;
        let pt_target = Point3::new(x0 + s * dir.x, y0 + s * dir.y, 0.0);
        let pt_camera = pose.transform_point(&pt_target);
        if let Some(proj) = camera.project_point(&pt_camera) {
            let pixel = Pt2::new(proj.x, proj.y);
            let ray = camera.backproject_pixel(&pixel);
            if !ray.point.x.is_finite() || !ray.point.y.is_finite() || !ray.point.z.is_finite() {
                continue;
            }
            if ray.point.z.abs() < 1e-6 {
                continue;
            }
            let ray_dir_camera = ray.point.normalize();
            let ray_dir_target = pose.inverse_transform_vector(&ray_dir_camera);
            if ray_dir_target.z.abs() > 1e-3 {
                pixels.push(pixel);
            }
        }
    }
    pixels
}

fn main() -> anyhow::Result<()> {
    // Ground truth camera (pinhole + Brown-Conrady + identity sensor)
    let intrinsics = FxFyCxCySkew {
        fx: 800.0,
        fy: 780.0,
        cx: 640.0,
        cy: 360.0,
        skew: 0.0,
    };
    let distortion = BrownConrady5 {
        k1: -0.2,
        k2: 0.05,
        k3: 0.0,
        p1: 0.001,
        p2: -0.001,
        iters: 8,
    };
    let sensor = ScheimpflugParams::default();
    let camera = Camera::new(Pinhole, distortion, sensor.compile(), intrinsics);

    // Synthetic dataset
    let points_3d = make_target_points();
    let poses = make_poses();
    let plane = make_laser_plane(&poses);

    let mut views: Vec<LaserlineView> = Vec::new();
    for pose in &poses {
        let mut points_2d = Vec::new();
        let mut points_3d_view = Vec::new();
        for pw in &points_3d {
            let p_cam = pose.transform_point(pw);
            if let Some(px) = camera.project_point(&p_cam) {
                points_3d_view.push(*pw);
                points_2d.push(Pt2::new(px.x, px.y));
            }
        }

        let laser_pixels = laser_pixels_for_view(pose, &plane, &camera);
        if points_2d.len() >= 4 && !laser_pixels.is_empty() {
            let obs = CorrespondenceView::new(points_3d_view, points_2d)?;
            views.push(View::new(
                obs,
                LaserlineMeta {
                    laser_pixels,
                    laser_weights: None,
                },
            ));
        }
    }

    let mut session = CalibrationSession::<LaserlineDeviceProblem>::new();
    session.set_input(views)?;

    run_calibration(&mut session, Some(LaserlineDeviceConfig::default()))?;
    let export = session.export()?;

    println!(
        "Calibrated intrinsics: fx={:.1}, fy={:.1}, cx={:.1}, cy={:.1}",
        export.estimate.params.intrinsics.fx,
        export.estimate.params.intrinsics.fy,
        export.estimate.params.intrinsics.cx,
        export.estimate.params.intrinsics.cy,
    );
    println!(
        "Mean reproj error: {:.3} px, mean laser error: {:.3}",
        export.stats.mean_reproj_error, export.stats.mean_laser_error
    );
    Ok(())
}

File: crates/vision-calibration-py/examples/laserline_device_session.py
"""Single laserline-device calibration with synthetic data.

Python counterpart of:
`crates/vision-calibration/examples/laserline_device_session.rs`
"""

from __future__ import annotations

import vision_calibration as vc

from _common import (
    add3,
    dot3,
    grid_points,
    make_pose,
    mat3_mul_vec,
    mat3_transpose,
    norm3,
    normalize3,
    pose_to_rt,
    project_point_pinhole,
    scale3,
    transform_point,
)


def make_poses() -> list[dict]:
    return [
        make_pose(0.0, 0.0, 0.0, 0.00, 0.00, 0.50),
        make_pose(0.15, -0.05, 0.0, 0.05, -0.02, 0.55),
        make_pose(-0.10, 0.08, 0.02, -0.04, 0.03, 0.60),
        make_pose(0.05, 0.12, -0.04, 0.02, 0.06, 0.52),
        make_pose(-0.12, -0.04, 0.06, -0.06, -0.04, 0.58),
    ]


def make_laser_plane(poses: list[dict]) -> dict:
    normal = normalize3([0.1, 0.05, 0.99])
    target_center = [0.075, 0.06, 0.0]
    center_cam = transform_point(poses[0], target_center)
    return {"normal": normal, "distance": -dot3(normal, center_cam)}


def laser_pixels_for_view(
    pose: dict,
    plane: dict,
    intrinsics: dict,
    distortion: dict,
) -> list[list[float]]:
    r, t = pose_to_rt(pose)
    rt = mat3_transpose(r)
    n_c = list(plane["normal"])
    n_t = mat3_mul_vec(rt, n_c)
    d_t = dot3(n_c, t) + float(plane["distance"])

    direction = [n_t[1], -n_t[0], 0.0]
    direction_norm = norm3(direction)
    if direction_norm < 1e-12:
        return []
    direction = scale3(direction, 1.0 / direction_norm)

    if abs(n_t[0]) > abs(n_t[1]):
        x0, y0 = -d_t / n_t[0], 0.0
    else:
        x0, y0 = 0.0, -d_t / n_t[1]

    pixels: list[list[float]] = []
    for i in range(40):
        s = (i / 39.0) * 0.2 - 0.1
        p_target = [x0 + s * direction[0], y0 + s * direction[1], 0.0]
        p_cam = transform_point(pose, p_target)
        uv = project_point_pinhole(p_cam, intrinsics, distortion)
        if uv is None:
            continue
        if not (uv[0] == uv[0] and uv[1] == uv[1]):
            continue
        if p_cam[2] <= 1e-6:
            continue
        ray_dir_cam = normalize3(p_cam)
        ray_dir_target = mat3_mul_vec(rt, ray_dir_cam)
        if abs(ray_dir_target[2]) <= 1e-3:
            continue
        if not (0.0 <= uv[0] < 1280.0 and 0.0 <= uv[1] < 720.0):
            continue
        pixels.append(uv)
    return pixels


def make_input() -> vc.LaserlineDataset:
    intrinsics = {"fx": 800.0, "fy": 780.0, "cx": 640.0, "cy": 360.0, "skew": 0.0}
    distortion = {"k1": 0.0, "k2": 0.0, "k3": 0.0, "p1": 0.0, "p2": 0.0, "iters": 8}
    points_3d = grid_points(cols=6, rows=5, spacing_m=0.03)
    poses = make_poses()
    plane = make_laser_plane(poses)

    views: list[vc.LaserlineView] = []
    for pose in poses:
        obs_3d: list[list[float]] = []
        obs_2d: list[list[float]] = []
        for p in points_3d:
            p_cam = transform_point(pose, p)
            uv = project_point_pinhole(p_cam, intrinsics, distortion)
            if uv is None:
                continue
            obs_3d.append(p)
            obs_2d.append(uv)

        laser_pixels = laser_pixels_for_view(pose, plane, intrinsics, distortion)
        if len(obs_2d) >= 4 and laser_pixels:
            views.append(
                vc.LaserlineView(
                    observation=vc.Observation(points_3d=obs_3d, points_2d=obs_2d),
                    laser_pixels=[tuple(p) for p in laser_pixels],
                    laser_weights=None,
                )
            )

    return vc.LaserlineDataset(views=views)


def main() -> None:
    print("=== Laserline Device Calibration (Synthetic, Python) ===")
    payload = make_input()
    result = vc.run_laserline_device(payload)
    intr = result.estimate["params"]["intrinsics"]
    plane = result.estimate["params"]["plane"]

    print(f"Views: {len(payload.views)}")
    print(f"Final cost: {result.estimate['report']['final_cost']:.3e}")
    print(f"Mean reprojection error: {result.mean_reproj_error:.6f} px")
    print(f"Mean laser error: {result.mean_laser_error:.6f}")
    print(f"Recovered intrinsics: fx={intr['fx']:.3f}, fy={intr['fy']:.3f}, cx={intr['cx']:.3f}, cy={intr['cy']:.3f}")
    print(f"Recovered plane: normal={plane['normal']}, distance={plane['distance']:.6f}")


if __name__ == "__main__":
    main()
